From dda5a323971655f314d5d464e3fd3ce34ea0f684 Mon Sep 17 00:00:00 2001
From: Leo Liang <lliang@marvell.com>
Date: Thu, 9 Feb 2017 14:26:50 +0800
Subject: [PATCH 20/45] fix: usb: gadget: clear usb descriptors pointers to
 avoid kernel crash

Issue:
 - When udc and u3d drivers enabled on Armada 3700 DB board, connecting
   device to PC host USB2 port, plug/unplug USB cable a few times,
   kernel crash occurs.

Issue description:
 - Both of udc and u3d drivers are intruduced by Marvell and switch them
   in glue layer in runtime. First, while USB cable is not connected,
   the fs/hs/ss descriptors are initialized. Then connectiong to PC USB2
   port, the glue layer will disconnect u3d driver and re-connect to udc
   driver. When u3d disconnected, all descripters are freed and not set
   to NULL in usb_free_all_descriptors function.. When re-connecting,
   gadget is running on high speed mode, the fs/hs descriptors are
   re-allocaled. But the ss descriptor is not changed at this time and
   still point to some object. If this pointer is re-allocaled by kernel
   for other system (for example, fs sub system), no crash at this time.
   But after unplug USB cable from PC, ths ss descriptor pointer will be
   freed again. If this pointer is not used by kernel, no problem and
   just a warning if slab debug enabled. That's why we need to plug or
   unplug several times for reproduce this issue. If this pointer is
   used by other system and will be crashed while the memory do read or
   write in next time.

Fix:
 - To avoid this problem, set all descriptors to NULL.

This patch fixes SYSTEMSW-3066 JIRA.

Change-Id: I773a5e2c275b6266c49c9fa171bab871692a91e0
Signed-off-by: Leo Liang <lliang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37293
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
---
 drivers/usb/gadget/config.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/usb/gadget/config.c b/drivers/usb/gadget/config.c
index 80a75c038696..7c4235b7ec64 100644
--- a/drivers/usb/gadget/config.c
+++ b/drivers/usb/gadget/config.c
@@ -199,8 +199,11 @@ EXPORT_SYMBOL_GPL(usb_assign_descriptors);
 void usb_free_all_descriptors(struct usb_function *f)
 {
 	usb_free_descriptors(f->fs_descriptors);
+	f->fs_descriptors = NULL;
 	usb_free_descriptors(f->hs_descriptors);
+	f->hs_descriptors = NULL;
 	usb_free_descriptors(f->ss_descriptors);
+	f->ss_descriptors = NULL;
 	usb_free_descriptors(f->ssp_descriptors);
 }
 EXPORT_SYMBOL_GPL(usb_free_all_descriptors);
-- 
2.35.3

